service/src/main/kotlin/fi/espoo/evaka/shared/job/ScheduledJobs.kt

```diff
@@ -0,0 +1,135 @@
+import fi.espoo.evaka.varda.VardaUpdateService
+import fi.espoo.voltti.logging.loggers.info
+import mu.KotlinLogging
+import org.springframework.stereotype.Component
+import java.time.Instant
+import java.time.LocalDate
+
+enum class ScheduledJob(val fn: (ScheduledJobs, Database.Connection) -> Unit) {
+    CancelOutdatedTransferApplications(ScheduledJobs::cancelOutdatedTransferApplications),
+    DvvUpdate(ScheduledJobs::dvvUpdate),
+    EndOfDayAttendanceUpkeep(ScheduledJobs::endOfDayAttendanceUpkeep),
+    EndOutdatedVoucherValueDecisions(ScheduledJobs::endOutdatedVoucherValueDecisions),
+    FreezeVoucherValueReports(ScheduledJobs::freezeVoucherValueReports),
+    KoskiUpdate(ScheduledJobs::koskiUpdate),
+    RemoveOldAsyncJobs(ScheduledJobs::removeOldAsyncJobs),
+    RemoveOldDaycareDailyNotes(ScheduledJobs::removeOldDaycareDailyNotes),
+    RemoveOldDraftApplications(ScheduledJobs::removeOldDraftApplications),
+    SendPendingDecisionReminderEmails(ScheduledJobs::sendPendingDecisionReminderEmails),
+    VardaUpdate(ScheduledJobs::vardaUpdate),
+}
+
+private val logger = KotlinLogging.logger { }
+
+@Component
+class ScheduledJobs(
+    private val vardaUpdateService: VardaUpdateService,
+    private val dvvModificationsBatchRefreshService: DvvModificationsBatchRefreshService,
+    private val attachmentsController: AttachmentsController,
+    private val pendingDecisionEmailService: PendingDecisionEmailService,
+    private val voucherValueDecisionService: VoucherValueDecisionService,
+    private val koskiUpdateService: KoskiUpdateService,
+    asyncJobRunner: AsyncJobRunner
+) {
+
+    init {
+        asyncJobRunner.runScheduledJob = { db, msg ->
+            val logMeta = mapOf("jobName" to msg.job.name)
+            logger.info(logMeta) { "Running scheduled job ${msg.job.name}" }
+            db.connect { msg.job.fn(this, it) }
+        }
+    }
+
+    fun endOfDayAttendanceUpkeep(db: Database.Connection) {
+        db.transaction {
+            it.createUpdate(
+                // language=SQL
+                """
+                    UPDATE child_attendance ca
+                    SET departed = ((arrived AT TIME ZONE 'Europe/Helsinki')::date + time '23:59') AT TIME ZONE 'Europe/Helsinki'
+                    WHERE departed IS NULL
+                """.trimIndent()
+            ).execute()
+        }
+    }
+
+    fun dvvUpdate(db: Database.Connection) {
+        dvvModificationsBatchRefreshService.scheduleBatch(db)
+    }
+
+    fun koskiUpdate(db: Database.Connection) {
+        koskiUpdateService.scheduleKoskiUploads(db, KoskiSearchParams())
+    }
+
+    fun vardaUpdate(db: Database.Connection) {
+        vardaUpdateService.scheduleVardaUpdate(db)
+    }
+
+    fun removeOldDraftApplications(db: Database.Connection) {
+        db.transaction { it.removeOldDrafts(attachmentsController::deleteAttachment) }
+    }
+
+    fun cancelOutdatedTransferApplications(db: Database.Connection) {
```

Comment:
Should the audit logging be placed here now that it won't be in the controller? Same for all of the jobs where auditing is relevant.

And just for posterity: audit logs are (in Voltti's case at least but also commonly) stored for much longer than other "technical" logs.

Reply:
I thought about this and decided to drop the audit logs, because

a) the jobs are nightly "batch" jobs, and not something that are started by humans. If we add a REST endpoint for manual triggering, we can of course add audit logging there
b) the old jobs were inconsistent: some had audit logging and some didn't

Reply:
By the way, I don't know if you noticed, but canceling outdated transfer applications is not on in production. I don't know if this PR enables it in prod, but that's something that should be taken into account.

Reply:
Good catch :cold_sweat: I didn't notice, because I just looked at the "schedule" variables in terraform and didn't notice many jobs had separate "enabled" variables.

I'll check all the jobs and will probably have to look at env variables in EspooConfig.kt

Reply:
Regarding audit logging: I think the only questionable one is `dvvUpdate` **but** I think we fulfill the VTJ requirements by having audit logging when users make requests for data that originated from VTJ:

> Luvan saaja vastaa siitä, että jälkikäteen on selvitettävissä henkilötasolla, kuka väestötietojärjestelmästä haettuja tietoja on saanut asiakasjärjestelmän kautta nähtäväkseen. Organisaatio, jolle oikeus sovelluskyselyyn on annettu, on velvollinen säilyttämään lokitiedot viideltä (5) vuodelta.

So on that part -> :heavy_check_mark: 

But good catch on the jobs :+1: